src/shell_manager.cc

```diff
@@ -325,6 +325,235 @@ std::pair<String, int> ShellManager::eval(
+    StringView cmdline, const Context& context, Vector<StringView> inputs,
+    Flags flags, const ShellContext& shell_context)
+{
+    const DebugFlags debug_flags = context.options()["debug"].get<DebugFlags>();
+    const bool profile = debug_flags & DebugFlags::Profile;
+    if (debug_flags & DebugFlags::Shell)
+        write_to_debug_buffer(format("shell:\n{}\n----\n", cmdline));
+
+    auto start_time = profile ? Clock::now() : Clock::time_point{};
+
+    auto kak_env = generate_env(cmdline, context, shell_context);
+
+    auto spawn_time = profile ? Clock::now() : Clock::time_point{};
+
+    struct PipeReader : FDWatcher
+    {
+        String contents;
+
+        PipeReader(Pipe& pipe)
+            : FDWatcher(pipe.read_fd(), FdEvents::Read,
+                        [this, &pipe](FDWatcher& watcher, FdEvents, EventMode) {
+                            char buffer[1024];
+                            while (fd_readable(pipe.read_fd()))
+                            {
+                                size_t size = ::read(pipe.read_fd(), buffer, 1024);
+                                if (size <= 0)
+                                {
+                                    pipe.close_read_fd();
+                                    watcher.disable();
+                                    return;
+                                }
+                                contents += StringView{buffer, buffer+size};
+                            }
+                        })
+        {}
+    };
+
+    struct PipeWriter : FDWatcher
+    {
+        PipeWriter(Pipe& pipe, StringView contents)
+            : FDWatcher(pipe.write_fd(), FdEvents::Write,
+                        [contents, &pipe](FDWatcher& watcher, FdEvents, EventMode) mutable {
+                            while (fd_writable(pipe.write_fd()))
+                            {
+                                ssize_t size = ::write(pipe.write_fd(), contents.begin(),
+                                                       (size_t)contents.length());
+                                if (size > 0)
+                                    contents = contents.substr(ByteCount{(int)size});
+                                if (size == -1 and (errno == EAGAIN or errno == EWOULDBLOCK))
+                                    return;
+                                if (size < 0 or contents.empty())
+                                {
+                                    pipe.close_write_fd();
+                                    watcher.disable();
+                                    return;
+                                }
+                            }
+                        })
+        {
+            int flags = fcntl(pipe.write_fd(), F_GETFL, 0);
+            fcntl(pipe.write_fd(), F_SETFL, flags | O_NONBLOCK);
+        }
+
+    };
+;    
+
+    struct PipeProcess
+    {
+        pid_t pid;
+        PipeWriter *stdin;
+        PipeReader *stdout, *stderr;
+        Pipe child_stdin, child_stdout, child_stderr;
```

Comment:
Move the Pipe objects before, so that they are constructed first, and use direct PipeWriter/PipeReader object instead of pointers, that will simplify logic here (no need to new/delete).

Reply:
The reason I went for this solution is because PipeWriter and PipeReader do not have a copy constructor. We might add this in a possible final implementation